[http1] Refactor H/1 codec with a parser interface

asraa opened the pull request:
Co-authored-by: Derek Argueta <darguetap@gmail.com>
Signed-off-by: Asra Ali <asraa@google.com>

Commit Message: Refactors the HTTP/1 codec to use an interface Parser. http_parser implements the Parser as the legacy parser (hard-coded in) to prepare for a new parser implementation with llhttp.
Risk Level: Medium. big refactor.
Testing: Unit and integration tests pass.
See https://github.com/envoyproxy/envoy/pull/9033 for history.
Part of: https://github.com/envoyproxy/envoy/issues/5155


Performance: there is a small performance degradation with the refactor https://gist.github.com/asraa/8babd3516acd447aa0f2ff5aa61d2e19


BTW on naming the methods. 
* Names like `onHeadersComplete()`, `onMessageComplete()` are top-level virtual methods in the Parser. Server/Client specific sub-routings are called `onHeadersCompleteBase()`

derekargueta commented on source/common/http/http1/legacy_parser_impl.cc:
```diff
@@ -0,0 +1,178 @@
+
+#include <http_parser.h>
+
+#include <iostream>
```
I think this is a stray include

derekargueta commented on source/common/http/http1/legacy_parser_impl.cc:
```diff
@@ -0,0 +1,178 @@
+        [](http_parser* parser, const char* at, size_t length) -> int {
+          return static_cast<ParserCallbacks*>(parser->data)->onUrl(at, length);
+        },
+        // TODO(dereka) onStatus
```
I think we can get rid of these TODO comments (`on_status` and `on_chunk_complete`). We don't use them today.

derekargueta commented on source/common/http/http1/codec_impl.h:
```diff
@@ -367,39 +353,42 @@ class ConnectionImpl : public virtual Connection,
    */
-  Status onHeaderField(const char* data, size_t length);
+  int onHeaderField(const char* data, size_t length) override;
+  Status onHeaderFieldStatus(const char* data, size_t length);
```
these
```cpp
int foo(const char* data, size_t length)
Status fooStatus(const char* data, size_t length)
```
methods were a bit of a quick hack for me to work around the newer `RETURN_IF_ERROR`/Status stuff, but I think we can do better. Since the callback functions return ints, we basically need a way to call `setAndCheckCallbackStatus` on the result of `RETURN_IF_ERROR` and return _that_, without which we get a type error trying to return a Status.

asraa replied:
Yep I agree, we essentially have extra functions defined. 

Ideally I think that the `int` parser-dependent status code should be localized to the parser code and codec_impl.cc shouldn't see them.

I wonder if I can do a `parser_->statusToInt(x)` in the parser code, but make `ParserStatus setAndCheckCallbackStatus` run at the end of `foo`?

derekargueta replied:
Best I can come with at the moment is maybe making `codec_status_` a part of the parser library rather than the codec library. Reading over the codec it _seems_ to primarily be more informational about the parser library, and might make more sense there with the codec calling something like `parser_->status().ok()`. That would enable to do `setAndCheckCallbackStatus` in the parser and have the `ParserCallbacks` methods just return a `Status` (thus the `RETURN_IF_ERROR` "just works"), keeping a lot of that flow the same as what we currently do with the http_parser hardcoded in.

So rather than teaching the codec to return `int`s for the `ParserCallback`, we teach the `ParserCallback` to accept `Status` and do the `setAndCheckCallbackStatus` conversion.

e.g. pre-PR
```cpp
// codec_impl.cc
    [](http_parser* parser) -> int {
      auto* conn_impl = static_cast<ConnectionImpl*>(parser->data);
      auto status = conn_impl->onMessageBeginBase();
      return conn_impl->setAndCheckCallbackStatus(std::move(status));
    },
```
would become
```cpp
// legacy_parser_impl.cc
        [](http_parser* parser) -> int {
          auto* conn_impl = static_cast<ParserCallback*>(parser->data);
          auto status = conn_impl->onMessageBegin();
          return conn_impl->setAndCheckCallbackStatus(std::move(status));
        },
```
It'd be a bigger mechanical change though

asraa replied:
Hmmm yeah! That works! The codec can call into the parser for the status. Will implement.

asraa replied:
YAY! all done.

derekargueta commented:
Thanks for getting this pushed Asra! 🚀

yanavlasov commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -614,10 +565,10 @@ Http::Status ConnectionImpl::innerDispatch(Buffer::Instance& data) {
         // Parse errors trigger an exception in dispatchSlice so we are guaranteed to be paused at
         // this point.
-        ASSERT(HTTP_PARSER_ERRNO(&parser_) == HPE_PAUSED);
+        ASSERT(parser_->getErrno() == parser_->statusToInt(ParserStatus::Paused));
```
If think this pattern is used in quite all other places. Would it make sense to add `Parser::getStatus()` method that translates parser's error code to status so that the code would look lie this:
```
ASSERT(parser_->getStatus() == ParserStatus::Paused);
``` 
I think it will make it more readable.

yanavlasov commented:
/wait

asraa commented:
Yay! Thanks Yan! 
@envoyproxy/senior-maintainers would love more eyes

mattklein123 reviewed (approved):
I just skimmed through and looks great to me! Not sure if @alyssawilk wants to take a quick look or not.

asraa commented:
(sorry about that -- self reviewed and saw an unused method in the header file I did not remove after Derek's review)

alyssawilk commented on source/common/http/http1/legacy_parser_impl.h:
```diff
@@ -0,0 +1,43 @@
+  bool seenContentLength() const override;
+  void setSeenContentLength(bool) override{};
+  int statusToInt(const ParserStatus code) const override;
+  int flagsChunked() const override { return 1; }
```
F_CHUNKED?  even if it is actually 1.

alyssawilk commented on source/common/http/http1/parser.h:
```diff
@@ -0,0 +1,183 @@
+
+enum class Method {
+  Head = 2,
+  Connect = 5,
```
Again should these use the http parser constants (cast if necessary)?
Also are they guaranteed to be consistent for parser next and parser legacy?  should they be functions as well?

asraa replied:
They (these Methods the codec actually cares about) are consistent between the new and old parser. We could actually eliminate these if the codec compared with method strings... that might be better than making the parser interface depend on the impl's constants

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -1097,24 +1052,25 @@ Envoy::StatusOr<ConnectionImpl::HttpParserCode> ServerConnectionImpl::onHeadersC
-    if (parser_.flags & F_CHUNKED ||
-        (parser_.content_length > 0 && parser_.content_length != ULLONG_MAX) || handling_upgrade_) {
+    if (parser_->flags() & parser_->flagsChunked() ||
+        (parser_->contentLength() > 0 && parser_->contentLength() != ULLONG_MAX) ||
```
lazy check: is the uulong_max a "not set" value?    if so is that going to be consistent between parsers or should content length be optional and the legacy version return nullopt if uulong_max?

edit: ah yeah, I think this is seenContentLength() && >0 but I'd mildly prefer an optional and explicit checks or a todo to convert if you want to minimize churn here.

asraa replied:
love the optional! and yes, the UULONG_MAX was an artifact where seenContentLength() should have been used.

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -1252,16 +1209,17 @@ Status ServerConnectionImpl::checkHeaderNameForUnderscores() {
-  } else if (parser_.status_code == 204 || parser_.status_code == 304 ||
-             (parser_.status_code >= 200 && parser_.content_length == 0 &&
-              !(parser_.flags & F_CHUNKED))) {
+  } else if (parser_->statusCode() == 204 || parser_->statusCode() == 304 ||
```
optional, do we want these returning the envoy status code values as long as we have an API?
I'm pro minimal churn but I'm also pro good APIs so I leave it up to you :-P

asraa replied:
Added a follow-up TODO to add in next PR

alyssawilk commented:
Happy to take a look -  I'm about halfway through but I'll ship whatever I have by EoD.

alyssawilk commented on source/common/http/http1/legacy_parser_impl.cc:
```diff
@@ -0,0 +1,209 @@
+    return ParserStatus::NoBody;
+  case 2:
+    return ParserStatus::NoBodyData;
+  case 31:
```
do we have links of where these magic numbers come from?

asraa replied:
https://github.com/envoyproxy/envoy/blob/4e71ed698356b68ca4f5ceca6b2fd128e88fe1ac/source/common/http/http1/legacy_parser_impl.cc#L190

Oops, it's linked in `statusToInt`, but not in `intToStatus`. Added

alyssawilk commented on source/common/http/http1/codec_impl.h:
```diff
@@ -238,31 +238,16 @@ class ConnectionImpl : public virtual Connection,
   bool resetStreamCalled() { return reset_stream_called_; }
-  Status onMessageBeginBase();
+
+  Status onMessageBegin() override;
```
comment on the difference between these two?

asraa replied:
Added a comment. `onMessageBeginStatus()` is the connection level method

alyssawilk commented on source/common/http/http1/codec_impl.h:
```diff
@@ -351,55 +336,19 @@ class ConnectionImpl : public virtual Connection,
-   * @param length supplies the length.
-   */
-  void bufferBody(const char* data, size_t length);
+  void bufferBody(const char* data, size_t length) override;
```
Should we collect these and comment what interface they're from?  TODO is fine if you're trying to minimize diffs here.

alyssawilk commented on source/common/http/http1/codec_impl.h:
```diff
@@ -418,17 +367,13 @@ class ConnectionImpl : public virtual Connection,
 
   /**
    * Called when accepting a chunk header.
    */
```
remove comment?

alyssawilk commented on source/common/http/http1/legacy_parser_impl.cc:
```diff
@@ -0,0 +1,209 @@
+  ParserStatus pause() {
+    http_parser_pause(&parser_, 1);
+    // http_parser pauses through http_parser_pause above.
+    return ParserStatus::Success;
```
is this return value in preparation for the new parser?

asraa replied:
Yep exactly. New parser doesn't pause through an API call but through a callback rval.

alyssawilk commented on source/common/http/http1/parser.h:
```diff
@@ -0,0 +1,183 @@
+  virtual uint64_t contentLength() const PURE;
+
+  // Returns parser flags (e.g. chunked).
+  virtual int flags() const PURE;
```
I'd be inclined to get rid of this and flagsChunked for a isChunked accessor.

alyssawilk commented on source/common/http/http1/parser.h:
```diff
@@ -0,0 +1,183 @@
+  virtual int flags() const PURE;
+
+  // Returns an integer representing the method. For requests only.
+  virtual uint16_t method() const PURE;
```
are these going to be consistent across parsers?  Better t ohave
isPostMethod() or isMethod(Http::Headers::get().MethodValues.Post)?

asraa replied:
Removed this in favor of returning a string and comparing with `MethodValues` per the other comment about the enum class. 

Still open to introduce `isMethod(_)`, which I like (we still need the method name in the codec)

alyssawilk commented on source/common/http/http1/parser.h:
```diff
@@ -0,0 +1,183 @@
+  virtual uint16_t method() const PURE;
+
+  // Returns a textual representation of the method. For requests only.
+  virtual const char* methodName() const PURE;
```
string view here and below?

alyssawilk commented on source/common/http/http1/legacy_parser_impl.cc:
```diff
@@ -0,0 +1,209 @@
+
+const char* LegacyHttpParserImpl::methodName() const { return impl_->methodName(); }
+
+const char* LegacyHttpParserImpl::errnoName() {
```
I'm not seeing where this is used - can we remove it?

alyssawilk commented on source/common/http/http1/parser.h:
```diff
@@ -0,0 +1,183 @@
+  virtual const char* errnoName(int rc) const PURE;
+
+  // Returns whether the Transfer-Encoding header is present.
+  virtual int usesTransferEncoding() const PURE;
```
preference for has over using

alyssawilk commented on source/common/http/http1/parser.h:
```diff
@@ -0,0 +1,183 @@
+  virtual bool seenContentLength() const PURE;
+
+  // Tells the parser that the Content-Length header is present.
+  virtual void setSeenContentLength(bool val) PURE;
```
is this needed?

asraa replied:
This is because the new parser doesn't have an "unset" content-length like http_parser. But I will remove this for now, and add that in with the new impl.

alyssawilk reviewed (commented):
Thanks for this massive refactor - exciting to see progress towards the new parser!

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -943,17 +942,16 @@ void ServerConnectionImpl::onEncodeComplete() {
   HeaderString path(Headers::get().Path);
 
-  bool is_connect = (method == static_cast<int>(Method::Connect));
+  bool is_connect = (method == "CONNECT");
```
Headers::get().MethodValues.Connect ?

alyssawilk commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -943,17 +942,16 @@ void ServerConnectionImpl::onEncodeComplete() {
       (active_request.request_url_.getStringView()[0] == '/' ||
-       ((method == static_cast<int>(Method::Options)) &&
-        active_request.request_url_.getStringView()[0] == '*'))) {
+       (method == "OPTIONS" && active_request.request_url_.getStringView()[0] == '*'))) {
```
Headers::get().MethodValues.Options

alyssawilk commented on source/common/http/http1/parser.h:
```diff
@@ -136,44 +137,23 @@ class Parser {
-
-  // Returns an integer representing the method. For requests only.
-  virtual uint16_t method() const PURE;
+  // Returns whether parser flags have chunked set.
```
optionally refer to headers rather than flags?

alyssawilk reviewed (commented):
Looking great! Just a couple more nits (and that just because I'm optimizing for neatness - Enovy uses string constants instead of header values all over and I'm probably the cause of half of them =P)

alyssawilk reviewed (approved):
Sweet!

asraa commented:
Having a really rough time with verify_examples (locally unreproducible). looks like grpc-bridge. Don't see flakes

phlax commented:
@asraa im seeing the issue described here - but im also seeing a different issue with the `grpc-bridge` example which i think is more to do with the pr needing to merge main (might be wrong) 

https://dev.azure.com/cncf/envoy/_build/results?buildId=69431&view=logs&j=b3b483e7-4801-520f-bbe4-cd01268aca64&t=36cbf9a1-17be-57b0-fb1c-6b7661dec3a4&l=6310

phlax commented:
hmm - i see main is merged - the other issue i think is unrelated to the ssl bug tho

phlax commented:
this is the issue

```
Invalid type URL, unknown type: envoy.extensions.access_loggers.stdoutput.v3.StdoutputAccessLog for type Any
```

asraa commented:
Nice, my merge didn't pick up the PR
https://github.com/envoyproxy/envoy/pull/15494

Thanks!

phlax commented:
ah k - so not a bug - i was going to open another flake issue - phew!

alyssawilk commented:
ah, so verify examples failures should all merge main?  Nice catch!
I'll tag the PRs I notice, other @envoyproxy/maintainers can do the same

jmarantz commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -742,7 +695,7 @@ Envoy::StatusOr<ConnectionImpl::HttpParserCode> ConnectionImpl::onHeadersComplet
     }
   }
-  if (parser_.method == HTTP_CONNECT) {
+  if (parser_->methodName() == Headers::get().MethodValues.Connect) {
```
minor perf/readability nit: capture out Headers::get() as a temp, here and elsewhere in this file.

asraa commented:
I think this is a legit failure in grpc-bridge, trying to debug with running locally, but I'll need to rebuild the docker images that verify_examples uses -- just turning up trace logging to identify the issue doesn't work (and grpc-bridge doesn't have integration tests, but maybe that's the fastest way to repro outside of verify_examples) @phlax

asraa commented:
@phlax I was trying to build my own envoy binary locally and try to re-run the grpc-bridge verify examples. at this point i don't think it's a flake, but I am still pretty stuck debugging it.

@zuercher @snowp do you have any idea what might be causing grpc-bridge to fail from the grpc-bridge side?

phlax commented:
@asraa is  only failing locally or in ci also ?

asraa commented:
I'm pretty sure it's local as well. I built my own envoy, spun up both proxies with the config, spun up the client and server, and ran the code, and couldn't get back the header per the example.

I couldn't run with the artifact created in CI or just be running ./verify_examples so I'm not 100% sure if something else went wrong.

phlax commented:
the easier way to test an individual example/sandbox (eg grpc) is:

```
cd examples/grpc-bridge
./verify.sh
```

i would avoid the verify_examples.sh script

phlax commented:
not sure what the cause of the fail is @asraa  - the fail is here 

https://dev.azure.com/cncf/envoy/_build/results?buildId=70003&view=logs&j=f1a49366-51c0-540f-0294-f73c2bb311a9&t=7fd036c9-61ec-5b59-3645-e90950dc67a6&l=6940

it doesnt look like a flake

asraa commented:
Yeah, I agree -- I'm trying right now to fix up the ci/Dockerfile-envoy to grab my locally build binary and change the dockerfiles in grpc bridge to use my local one, turn up trace logging and check

phlax commented:
if you build the docker image with your binary and then set

```
export DOCKER_NO_PULL=1
```

as long as your docker image is correctly named it should use yours without attempting to pull the one from dockerhub

asraa commented:
Sounds good, I think that's what I have, I've built envoy (and su-exec) binary, moved it to build_envoy/envoy-binary, run 
`docker build -f ci/Dockerfile-envoy -t envoy .`
to get envoy:latest, 

and changed grpc-bridge/Dockerfile-(server/client) to pull `FROM envoy:latest`

Seems like envoy is starting up, but I get failures when sending the request:
```
> [grpc-bridge] Set key value foo=bar
Traceback (most recent call last):
  File "/client/grpc-kv-client.py", line 85, in <module>
    run()
  File "/client/grpc-kv-client.py", line 79, in run
    response = client.set(key, value)
  File "/client/grpc-kv-client.py", line 40, in set
    return requests.post(HOST + "/kv.KV/Set", data=data, headers=HEADERS)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 119, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', BadStatusLine("''",))
ERROR: 1
```

asraa commented:
Got it!! I followed the directs in examples/grpc-bridge to generate the stubs and bring everything up. I see the failure in the parser on an ASSERT. Thanks!

mattklein123 commented:
Great news that you figured this out. @asraa would you mind adding a dedicated test for whatever broke that doesn't depend on verify examples?

/wait

mattklein123 reviewed (approved):
Thank you!
